ext/bcmath/libbcmath/src/convert.c

```diff
@@ -0,0 +1,59 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Niels Dossche <nielsdos@php.net>                            |
+   +----------------------------------------------------------------------+
+*/
+
+#include "bcmath.h"
+#include "convert.h"
+
+#define SWAR_ONES (~((size_t) 0) / 0xFF)
+#define SWAR_REPEAT(x) (SWAR_ONES * (x))
+
+static char *bc_copy_and_shift_numbers(char *dest, const char *source, const char *source_end, unsigned char shift, bool add)
```

Comment:
Possible add a `restrict` keyword for `dest` and `source`? As those APIs are only used within other C files, so we don't need to have compatibility with C++?

Reply:
Right, I'll add the restrict keyword.

I'll explain what the code does.
The intention is to copy each byte from source to dest, but subtract or add '0' to each byte.
The idea of this patch is to try to read+write 8 bytes at once, adding/subtracting '0' to each byte also in parallel.

SWAR_ONES will be of the form 0x01010101 for 32-bit or 0x0101010101010101 for 64-bit.
Example: SWAR_ONES * 0xAB will therefore be equal to 0xABABABAB for 32-bit or 0xABABABABABABABAB for 64-bit.
So in this case, for SWAR_REPEAT('0'), it will be a 32/64-bit word where each byte is equal to '0', i.e. 0x303030...

Since we know that subtract/add overflow from one byte to another can't occur, we can subtract/add with 0x303030... to the entire 4/8 bytes which will be equivalent to adding 0x30 to each byte individually.

And to be complete: SWAR stands for "SIMD Within A Register"

Reply:
Ahhh okay, well could you please write this in a comment in the file? :D